bolt/lib/Profile/YAMLProfileReader.cpp

```diff
@@ -592,72 +633,276 @@ size_t YAMLProfileReader::matchWithCallGraph(BinaryContext &BC) {
+                         InlineSiteHash>;
+  using CallSiteInfoTy =
+      std::unordered_map<InlineSite, const CallSiteInfo *, InlineSiteHash>;
+  // Mapping from a parent node id to a map InlineSite -> Child node.
+  DenseMap<uint32_t, ChildMapTy> ParentToChildren;
+  // Collect calls in the profile: map from a parent node id to a map
+  // InlineSite -> CallSiteInfo ptr.
+  DenseMap<uint32_t, CallSiteInfoTy> ParentToCSI;
+  for (const BinaryBasicBlockProfile &YamlBB : YamlBF.Blocks) {
+    // Collect callees for inlined profile matching, indexed by InlineSite.
+    for (const CallSiteInfo &CSI : YamlBB.CallSites) {
+      ProbeMatchingStats.TotalCallCount += CSI.Count;
+      ++ProbeMatchingStats.TotalCallSites;
+      if (CSI.Probe == 0) {
+        LLVM_DEBUG(dbgs() << "no probe for " << CSI.DestId << " " << CSI.Count
+                          << '\n');
+        ++ProbeMatchingStats.MissingCallProbe;
+        ProbeMatchingStats.MissingCallCount += CSI.Count;
+        continue;
+      }
+      const BinaryFunctionProfile *Callee = IdToYamLBF.lookup(CSI.DestId);
+      if (!Callee) {
+        LLVM_DEBUG(dbgs() << "no callee for " << CSI.DestId << " " << CSI.Count
+                          << '\n');
+        ++ProbeMatchingStats.MissingCallee;
+        ProbeMatchingStats.MissingCallCount += CSI.Count;
+        continue;
+      }
+      // Get callee GUID
+      if (Callee->InlineTree.empty()) {
+        LLVM_DEBUG(dbgs() << "no inline tree for " << Callee->Name << '\n');
+        ++ProbeMatchingStats.MissingInlineTree;
+        ProbeMatchingStats.MissingCallCount += CSI.Count;
+        continue;
+      }
+      uint64_t CalleeGUID = Callee->InlineTree.front().GUID;
+      ParentToCSI[CSI.InlineTreeNode][InlineSite(CalleeGUID, CSI.Probe)] = &CSI;
+    }
+  }
+  LLVM_DEBUG({
+    for (auto &[ParentId, InlineSiteCSI] : ParentToCSI) {
+      for (auto &[InlineSite, CSI] : InlineSiteCSI) {
+        auto [CalleeGUID, CallSite] = InlineSite;
+        errs() << ParentId << "@" << CallSite << "->"
+               << Twine::utohexstr(CalleeGUID) << ": " << CSI->Count << ", "
+               << Twine::utohexstr(CSI->Offset) << '\n';
+      }
+    }
+  });
+
+  assert(!Root.isRoot());
+  LLVM_DEBUG(dbgs() << "matchInlineTreesImpl for " << BF << "@"
+                    << Twine::utohexstr(Root.Guid) << " and " << YamlBF.Name
+                    << "@" << Twine::utohexstr(FuncNode.GUID) << '\n');
+  ++ProbeMatchingStats.AttemptedNodes;
+  ++ProbeMatchingStats.AttemptedRoots;
+
+  // Match profile function with a lead node (top-level function or inlinee)
+  if (Root.Guid != FuncNode.GUID) {
+    LLVM_DEBUG(dbgs() << "Mismatching root GUID\n");
+    ++ProbeMatchingStats.MismatchingRootGUID;
+    return 0;
+  }
+  {
+    uint64_t BinaryHash = Decoder.getFuncDescForGUID(Root.Guid)->FuncHash;
+    uint64_t ProfileHash = FuncNode.Hash;
+    if (BinaryHash != ProfileHash) {
+      LLVM_DEBUG(dbgs() << "Mismatching hashes: "
```

Comment:
```suggestion
      LLVM_DEBUG(dbgs() << "BOLT-DEBUG: mismatching hashes: "
```

and more cases below.